docs/src/tutorial.md

```diff
@@ -31,62 +32,93 @@ An example of a space is [`OpenStreetMapSpace`](@ref). It is based on Open Stree
+    end
+    agent.age += 1
+    # if too old, pass fortune onto heir and remove from model
+    if agent.age > 75
+        heir = replicate!(agent, model)
+        heir.age = 1
+        remove_agent!(agent, modeL)
+    end
+    return # function end. As it is in-place it `return`s nothing.
+end
 ```
+and to activate all agents randomly once per simulation step, one would use `Schedulers.Randomly()` as the model scheduler.
 
-### Advanced stepping
-!!! note "Current step number"
-    Notice that the current step number is not explicitly given to the `model_step!`
-    function, because this is useful only for a subset of ABMs. If you need the
-    step information, implement this by adding a counting parameter into the model
-    `properties`, and incrementing it by 1 each time `model_step!` is called.
-    An example can be seen in the `model_step!` function of [Daisyworld](@ref),
-    where a `tick` is increased at each step.
+we stress that in contrast to the above `model_step!`, this function will be called for _every_ scheduled agent, while `model_step!` will only be called _once_ per simulation step.
+Naturally, you may define **both** an agent- and a model- stepping functions. In this case the model stepping function would perform model-wide actions that are not limited to a particular agent.
 
-The interface of [`step!`](@ref), which allows the option of both `agent_step!` and `model_step!` is driven mostly by convenience. In principle, the `model_step!` function by itself can perform all operations related with stepping the ABM.
-However, for many models, this simplified approach offers the benefit of not having to write an explicit loop over existing agents inside the `model_step!`.
-Most of the examples in our documentation can be expressed using an independent `agent_step!` and `model_step!` function.
+## [3.3 The evolution rule - advanced (manual scheduling)](@id manual_scheduling)
 
-On the other hand, more advanced models require special handling for scheduling, or may need to schedule several times and act on different subsets of agents with different functions.
-In such a scenario, it is more sensible to provide only a `model_step!` function (and use `dummystep` as `agent_step!`), where all configuration is contained within.
-Notice that if you follow this road, the argument `scheduler` given to [`AgentBasedModel`](@ref) somewhat loses its meaning.
+Some advanced models may require special handling for scheduling, or may need to schedule agents several times and act on different subsets of agents with different functions during a single simulation step.
+In such a scenario, it is more sensible to provide only a model stepping function,  where all configuration is contained within.
 
 Here is an example:
+
 ```julia
 function complex_step!(model)
+    scheduler1 = Schedulers.Randomly()
+    scheduler2 = user_defined_function
     for id in scheduler1(model)
         agent_step1!(model[id], model)
     end
```

Comment:
```suggestion
    scheduler1 = Schedulers.Randomly()
    scheduler2 = user_defined_function_with_model_as_arg
    for id in schedule(model, scheduler1)
        agent_step1!(model[id], model)
    end
```

same for `scheduler2`

Reply:
Why complicate things? A scheduler is a function that takes in the model as an argument. `schedule(model, scheduler)` has no reason to exist. It should be `scheduler(model)` instead. The only thing that should exist is `schedule(model)` because it uses the default internal scheduler.

Reply:
The problem this tries to address is that this way the check about removed agents is done automatically inside `schedule`, this seems to me a good safety reason to require that

Reply:
this is the code of `schedule(model, scheduler)`: https://github.com/JuliaDynamics/Agents.jl/pull/889/files#diff-8d4e4a89381c499f8748e691a16bd597edc7e55a035a72755fe69d4db4b0749bR13

Reply:
oooooooooooooh, now i understand. unfortunately this doesn't really help:
```julia
function schedule(model::ABM)
    scheduler = abmscheduler(model)(model)
    return Iterators.filter(id -> id in allids(model), scheduler)
end
schedule(model::UnremovableABM) = abmscheduler(model)(model)
schedule(model::ABM, scheduler) = Iterators.filter(id -> id in allids(model), scheduler(model))
```
You see, `allids(model)` is not dynamic. The only way to dynamically skip agents that are removed while the model evolves is the way we do it now:
```julia
for id in activation_order
        if model isa StandardABM # this if gets compiled out
            id in allids(model) || continue
        end
        agent_step!(model[id], model)
    end
```

Reply:
So `schedule` needs to be removed after all. It doens't add anything besides more `api` names.

`schedule` should be moved to the deprecations and be replaced by `abmscheduler(model)(model)`.

Reply:
Okay, forget what I said. `allids(model)` **is dynamic**.

Reply:
so over all your solution is fine, we keep `schedule` as is. OPlease adjust the source ccode accordingly.

Reply:
👍